File: plugins/agentic-skills/skills/workbench-builder/pr-workbench/scripts/review_loop.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["httpx"]
# ///
"""The terminal half of the review loop.

In a real session Claude Code runs this (or its logic inline): pull the queue of
things the human asked, do the work — read the diff, trace the import, draft the
comment — then POST a markdown response back. The browser updates live via SSE.

    uv run scripts/review_loop.py            # pull queue, auto-answer (demo)
    uv run scripts/review_loop.py --list     # just show what's queued
"""
import sys
import httpx

BASE = "http://127.0.0.1:5051"


def main() -> None:
    pull = httpx.get(f"{BASE}/claude/queue").json()
    reqs = pull["requests"]
    if not reqs:
        print("queue empty.")
        return
    if "--list" in sys.argv:
        for r in reqs:
            print(f"  #{r.get('pr_number','?')} [{r['kind']}] {r['body'][:70]}")
        return
    for r in reqs:
        # A real run would synthesize this from the actual diff/code. Demo replies:
        answer = _demo_answer(r)
        httpx.post(f"{BASE}/claude/respond", json={"request_id": r["id"], "response": answer})
        print(f"answered request {r['id']} on PR #{r.get('pr_number','?')}")


def _demo_answer(r: dict) -> str:
    kind, pr = r["kind"], r.get("pr_number", "?")
    if kind == "merge-check":
        return (f"**Not yet — merge #412 first.** PR #{pr} edits `app/models.py` and "
                f"`api/routes.py`, both of which #412 rewrites. Recommended order:\n\n"
                "1. `#412` (the extraction — everything rebases on it)\n"
                "2. `#415`, then `#419`\n3. `#417` last (smallest config touch)\n\n"
                "Rebase #" f"{pr} after #412 lands and re-run the suite.")
    if kind == "draft-comment":
        return ("> **Requesting changes.** Solid direction. Two things before approve:\n>\n"
                "> 1. The circular import (`billing.service` ↔ `app.models`) needs a protocol seam.\n"
                "> 2. Add the Alembic migration for the dropped `invoices.legacy_id` column.\n>\n"
                "> Happy to re-review once those land.")
    if kind == "investigate":
        return ("Traced it. The cycle is:\n\n```\napp/models.py  -> imports billing.service (new)\n"
                "billing/service.py -> imports app.models (Invoice)\n```\n\n"
                "Fix: define an `InvoiceLike` Protocol in `billing/types.py` and type against that, "
                "so `billing` never imports `app`. ~15 lines, no behavior change.")
    if kind == "summarize":
        return f"**PR #{pr} in one line:** moves invoice/subscription logic into a standalone `billing/` package; everything else in the set rebases on it."
    return f"Looked into PR #{pr}. {r['body']}\n\n*(demo response — a real run would answer from the diff.)*"


if __name__ == "__main__":
    main()

File: plugins/agentic-skills/skills/workbench-builder/pr-workbench/scripts/review-loop.ts
/**
 * Answer the human's review questions.
 *
 * Pulling /claude/queue CLAIMS requests (queued → working), so the badge flips
 * in the browser the moment this runs — the human sees pickup before the
 * answer exists. Replace the canned answer with real work: read the regions,
 * inspect the diff, then respond.
 *
 * Zero dependencies: global fetch.
 * Run: bun run scripts/review-loop.ts          (drain once and exit)
 *      bun run scripts/review-loop.ts --watch  (poll every 3s)
 */
const BASE = "http://127.0.0.1:5051";

type Req = {
  id: number;
  kind: string;
  body: string;
  pr_id: number | null;
  pr_number: number | null;
};

type Collision = { path: string; n: number; churn: number; prs: string };

async function get(path: string): Promise<any> {
  const r = await fetch(`${BASE}${path}`);
  if (!r.ok) throw new Error(`${path} → ${r.status}`);
  return r.json();
}

async function post(path: string, body: unknown): Promise<any> {
  const r = await fetch(`${BASE}${path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(`${path} → ${r.status} ${await r.text()}`);
  return r.json();
}

/**
 * Ground the answer in the room's own state instead of restating the question.
 * A per-PR ask gets that PR's overlap; a whole-set ask gets the contested
 * files — the same data the human is looking at, which is what makes the
 * answer checkable.
 */
async function answer(req: Req): Promise<string> {
  const collisions = (await get("/claude/collisions")) as Collision[];

  if (req.pr_id !== null) {
    const detail = await get(`/api/regions/pr-${req.pr_id}`);
    const contested: string[] = [
      ...new Set((detail.overlap as Array<{ path: string }>).map((o) => o.path)),
    ];
    const blockers = (detail.concerns as Array<{ severity: string; resolved: number }>).filter(
      (c) => c.severity === "blocker" && c.resolved === 0,
    ).length;
    return (
      `#${detail.pr.number}: ${detail.files.length} files, ${blockers} open blocker(s). `
      + (contested.length > 0
        ? `Contested with other PRs on ${contested.join(", ")} — rebase risk is there, not in the new files.`
        : "No file overlap with the rest of the set; it can land independently.")
    );
  }

  const hottest = collisions[0];
  return (
    `${collisions.length} contested file(s) across the set. `
    + (hottest
      ? `Hottest is ${hottest.path} (${hottest.n} PRs, ${hottest.churn} churn) — `
        + `land the smallest of those first so the rest rebase onto it.`
      : "No overlap: merge order is free.")
  );
}

async function drain(): Promise<number> {
  const { requests } = (await get("/claude/queue")) as { requests: Req[] };
  for (const req of requests) {
    console.log(`request #${req.id} (${req.kind}): ${req.body}`);
    await post("/claude/respond", { request_id: req.id, response: await answer(req) });
    console.log(`  answered #${req.id}`);
  }
  return requests.length;
}

if (process.argv.includes("--watch")) {
  console.log(`watching ${BASE}/claude/queue — ctrl-c to stop`);
  while (true) {
    await drain();
    await new Promise((res) => setTimeout(res, 3000));
  }
} else {
  const n = await drain();
  console.log(n === 0 ? "queue empty" : `drained ${n} request(s)`);
}
